src/test/java/org/opensearch/neuralsearch/common/BaseNeuralSearchIT.java

```diff
@@ -0,0 +1,164 @@
+
+import static org.apache.http.entity.ContentType.APPLICATION_JSON;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.function.Predicate;
+
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpHeaders;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.message.BasicHeader;
+import org.apache.http.util.EntityUtils;
+import org.opensearch.client.*;
+import org.opensearch.test.rest.OpenSearchRestTestCase;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableList;
+
+public abstract class BaseNeuralSearchIT extends OpenSearchRestTestCase {
+
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    private static final Locale locale = Locale.getDefault();
+
+    protected static final String pipelineName = "pipeline-hybrid";
+
+    protected final ClassLoader classLoader = this.getClass().getClassLoader();
+
+    public String uploadModel(String requestBody) throws Exception {
+        Response uploadResponse = makeRequest(
+            client(),
+            "POST",
+            "/_plugins/_ml/models/_upload",
+            null,
+            toHttpEntity(requestBody),
+            ImmutableList.of(new BasicHeader(HttpHeaders.USER_AGENT, "Kibana"))
+        );
+        JsonNode uploadResJson = objectMapper.readTree(EntityUtils.toString(uploadResponse.getEntity()));
+        String taskId = uploadResJson.get("task_id").asText();
+        assertNotNull(taskId);
+
+        JsonNode taskQueryResult = getTaskQueryResponse(taskId);
+        boolean isComplete = checkComplete(taskQueryResult);
+        while (!isComplete) {
+            taskQueryResult = getTaskQueryResponse(taskId);
+            isComplete = checkComplete(taskQueryResult);
+        }
+        String modelId = taskQueryResult.get("model_id").asText();
+        assertNotNull(modelId);
+        return modelId;
+    }
+
+    public void loadModel(String modelId) throws IOException {
+        Response uploadResponse = makeRequest(
+            client(),
+            "POST",
+            String.format(locale, "/_plugins/_ml/models/%s/_load", modelId),
+            null,
+            toHttpEntity(""),
+            ImmutableList.of(new BasicHeader(HttpHeaders.USER_AGENT, "Kibana"))
+        );
+        JsonNode uploadResJson = objectMapper.readTree(EntityUtils.toString(uploadResponse.getEntity()));
+        String taskId = uploadResJson.get("task_id").asText();
+        assertNotNull(taskId);
+
+        JsonNode taskQueryResult = getTaskQueryResponse(taskId);
+        boolean isComplete = checkComplete(taskQueryResult);
+        while (!isComplete) {
+            taskQueryResult = getTaskQueryResponse(taskId);
+            isComplete = checkComplete(taskQueryResult);
```

Comment:
This is a very brute force way of checking if model is uploaded or not. This is very resource intensive plus there is no time out which makes can make it run forever and just increase the build time.

Suggestion:
1. We should move this checking of model upload to another thread.
2. Add a timeout(3 times than the actual time, or let say 1 min) after which we will fail the test saying that model is not uploaded, and provide proper response why the tests failed and add things like increase the time out and other things.
3. To avoid the resource intensive work, we should also provide some sleep time in the thread which is checking the model upload if we apply \#1 suggestion.

Reply:
Add max retry time and sleep time in the thread.